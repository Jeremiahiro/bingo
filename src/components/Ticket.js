import React from 'react';
import PropTypes from 'prop-types';
import { Flex } from 'reflexbox';
import TicketRow from './TicketRow';

import './styles/Ticket.css';

/**
 * Main ticket component - renders the number rows on each ticket (TicketRow
 * component) as well as the prize indicator.
 * prize indicator beside
 * @class
 */
class Ticket extends React.Component {
  constructor(){
    super();
    this.state = {
      numbersToGo: 5,
      line: false,
      bingo: false
    }
  }

  componentWillReceiveProps(){
    let rowsToGo = 3;
    for (let row = 0; row < this.props.ticket.length; row++) {
      let numbersToGo = 5;
      for (let num = 0; num < this.props.ticket[row].length; num++) {
        let checkNumber = this.props.ticket[row][num];
        if (this.props.calledNumbers.includes(parseInt(checkNumber, 10))) {
          numbersToGo -= 1;
        }
      }
      if (numbersToGo === 0) {
        rowsToGo -= 1;
      }
      this._updateNumbersToGo(numbersToGo);
    }

    if (rowsToGo === 0){
      let state = Object.assign({}, this.state);
      state.bingo = true;
      this.setState(state);
    }

  }

  render() {

    let prizeText = null;
    if (this.state.bingo) {
      prizeText = <span><span className='bingo'>BINGO</span></span>;
    } else if (this.state.line) {
      prizeText = <span><span className='line'>LINE</span></span>;
    } else {
      prizeText = <span><span className='number'>{this.state.numbersToGo}</span> TO GO </span>
    }

    return (
      <Flex className="ticket-wrapper">
        <div className="number-wrapper">
            {this.props.ticket.map(function(row, index){
              return (
                <TicketRow
                  key={index}
                  row={row}
                  calledNumbers={this.props.calledNumbers}
                  />
              );
            }, this)}
        </div>
        <div className='prize-indicator'>
          {prizeText}
        </div>
      </Flex>
    );
  }

  _updateNumbersToGo(newVal){
    let state = Object.assign({}, this.state);
    if (newVal === 0) {
      state.line = true;
      state.numbersToGo = newVal;
      this.setState(state);
    } else if (newVal < this.state.numbersToGo) {
      state.numbersToGo = newVal;
      this.setState(state);
    }
  }
}

Ticket.propTypes = {
  ticket: PropTypes.array,
  calledNumbers: PropTypes.array,
};

export default Ticket;
