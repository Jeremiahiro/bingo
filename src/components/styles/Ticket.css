.number-wrapper {
  background-color: #ffdce5;
  padding: 10px;
  margin: 0;
}

div.prize-indicator {
  width: 40px;
  border: 2px black solid;
  margin: 10px 10px 8px 0;
  border-radius: 10px;
  background: linear-gradient(to bottom, #6bb56f 0%, #1b8e44 50%, #007d3f 51%, #186331 100%);
}

div.prize-indicator > span {
  color: #fff;
  display: table-cell;
  word-wrap: break-word;
  vertical-align: middle;
  word-spacing: 40px;
  line-height: 24px;
  padding-top: 6px;
}

div.prize-indicator > span > span.number {
  font-size: 30px;
  font-weight: bold;
}
