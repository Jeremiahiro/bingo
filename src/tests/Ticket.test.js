import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import Ticket from '../components/Ticket';

const ticket = [
  ['1', '', '2', '3', '', '', '4', '5'],
  ['6', '7', '8', '9', '10', '', '', ''],
  ['', '', '11', '12', '13', '14', '15', '']
];

const calledNumbersEmpty = [];
const calledNumbers      = [1, 2, 3];

it('renders without crashing when calledNumers is empty', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Ticket ticket={ticket} calledNumbers={calledNumbersEmpty}/>, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders without crashing when calledNumers has values', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Ticket ticket={ticket} calledNumbers={calledNumbers}/>, div);
  ReactDOM.unmountComponentAtNode(div);
});


it('renders the expected element when calledNumers is empty', () => {
  const component = renderer.create(
    <Ticket ticket={ticket} calledNumbers={calledNumbersEmpty}/>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders the expected element when calledNumers has values', () => {
  const component = renderer.create(
    <Ticket ticket={ticket} calledNumbers={calledNumbers}/>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
