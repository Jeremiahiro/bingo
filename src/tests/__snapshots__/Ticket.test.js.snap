// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the expected element when calledNumers has values 1`] = `
<div
  className="ticket-wrapper _Rfx0_"
>
  <div
    className="number-wrapper"
  >
    <div
      className="_Rfx0_"
    >
      <div
        className="ticket-number called _Rfx1_"
      >
        <span>
          1
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
      <div
        className="ticket-number called _Rfx1_"
      >
        <span>
          2
        </span>
      </div>
      <div
        className="ticket-number called _Rfx1_"
      >
        <span>
          3
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          4
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          5
        </span>
      </div>
    </div>
    <div
      className="_Rfx0_"
    >
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          6
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          7
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          8
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          9
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          10
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
    </div>
    <div
      className="_Rfx0_"
    >
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          11
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          12
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          13
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          14
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          15
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
    </div>
  </div>
  <div
    className="prize-indicator"
  >
    <span>
      <span
        className="number"
      >
        5
      </span>
       TO GO 
    </span>
  </div>
</div>
`;

exports[`renders the expected element when calledNumers is empty 1`] = `
<div
  className="ticket-wrapper _Rfx0_"
>
  <div
    className="number-wrapper"
  >
    <div
      className="_Rfx0_"
    >
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          1
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          2
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          3
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          4
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          5
        </span>
      </div>
    </div>
    <div
      className="_Rfx0_"
    >
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          6
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          7
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          8
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          9
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          10
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
    </div>
    <div
      className="_Rfx0_"
    >
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          11
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          12
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          13
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          14
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          15
        </span>
      </div>
      <div
        className="ticket-number  _Rfx1_"
      >
        <span>
          
        </span>
      </div>
    </div>
  </div>
  <div
    className="prize-indicator"
  >
    <span>
      <span
        className="number"
      >
        5
      </span>
       TO GO 
    </span>
  </div>
</div>
`;
